url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/recursive/ReliableDNSClient.java

```diff
@@ -0,0 +1,180 @@
+ */
+package de.measite.minidns.recursive;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.logging.Level;
+
+import de.measite.minidns.AbstractDNSClient;
+import de.measite.minidns.DNSCache;
+import de.measite.minidns.DNSClient;
+import de.measite.minidns.DNSMessage;
+import de.measite.minidns.Question;
+import de.measite.minidns.source.DNSDataSource;
+import de.measite.minidns.util.MultipleIoException;
+
+/**
+ * A DNS client using a reliable strategy. First the configured resolver of the
+ * system are used, then, in case there is no answer, a fall back to iterative
+ * resolving is performed.
+ */
+public class ReliableDNSClient extends AbstractDNSClient {
+
+    public enum Mode {
+        /**
+         * Try the recursive servers first and fallback to iterative resolving if it fails. This is the default mode.
+         */
+        recursiveWithIterativeFallback,
+
+        /**
+         * Only try the recursive servers. This makes {@code ReliableDNSClient} behave like a {@link DNSClient}.
+         */
+        recursiveOnly,
+
+        /**
+         * Only use iterative resolving.  This makes {@code ReliableDNSClient} behave like a {@link RecursiveDNSClient}.
+         */
+        iterativeOnly,
+    }
+
+    private final RecursiveDNSClient recursiveDnsClient;
+    private final DNSClient dnsClient;
+
+    private Mode mode = Mode.recursiveWithIterativeFallback;
+
+    public ReliableDNSClient(DNSCache dnsCache) {
+        super(dnsCache);
+        recursiveDnsClient = new RecursiveDNSClient(dnsCache) {
+            @Override
+            protected DNSMessage.Builder newQuestion(DNSMessage.Builder questionMessage) {
+                questionMessage = super.newQuestion(questionMessage);
+                return ReliableDNSClient.this.newQuestion(questionMessage);
+            }
+            @Override
+            protected boolean isResponseCacheable(Question q, DNSMessage dnsMessage) {
+                boolean res = super.isResponseCacheable(q, dnsMessage);
+                return ReliableDNSClient.this.isResponseCacheable(q, dnsMessage) && res;
+            }
+        };
+        dnsClient = new DNSClient(dnsCache) {
+            @Override
+            protected DNSMessage.Builder newQuestion(DNSMessage.Builder questionMessage) {
+                questionMessage = super.newQuestion(questionMessage);
+                return ReliableDNSClient.this.newQuestion(questionMessage);
+            }
+            @Override
+            protected boolean isResponseCacheable(Question q, DNSMessage dnsMessage) {
+                boolean res = super.isResponseCacheable(q, dnsMessage);
+                return ReliableDNSClient.this.isResponseCacheable(q, dnsMessage) && res;
+            }
+        };
+    }
+
+    public ReliableDNSClient() {
+        this(DEFAULT_CACHE);
+    }
+
+    @Override
+    protected DNSMessage query(DNSMessage.Builder q) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 93-125, lines 127-137 into separate methods.